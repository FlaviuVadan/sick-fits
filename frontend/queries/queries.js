import gql from 'graphql-tag';

import { perPage } from '../config';

export const ALL_ITEMS_QUERY = gql`
    query ALL_ITEMS_QUERY(
    $skip: Int = 0
    $first: Int = ${ perPage }
    ) {
        items(
            orderBy: createdAt_DESC
            skip: $skip
            first: $first
        ) {
            id
            title
            price
            description
            image
            largeImage
        }
    }
`;

export const CREATE_ITEM_MUTATION = gql`
    mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
    ) {
        createItem(
            title: $title
            description: $description
            price: $price
            image: $image
            largeImage: $largeImage
        ) {
            id
        }
    }
`;

export const UPDATE_ITEM_MUTATION = gql`
    mutation UPDATE_ITEM_MUTATION(
    $id: ID!
    $title: String
    $description: String
    $price: Int
    ) {
        updateItem(
            id: $id
            title: $title
            description: $description
            price: $price
        ) {
            id
            title
            description
            price
        }
    }
`;

export const ITEM_QUERY = gql`
    query ITEM_QUERY(
    $id: ID!
    ) {
        item(where: {
            id: $id
        }) {
            id
            title
            description
            price
            largeImage
        }
    }
`;

export const DELETE_ITEM_MUTATION = gql`
    mutation DELETE_ITEM_MUTATION(
    $id: ID!
    ) {
        deleteItem(
            id: $id
        ) {
            id
        }
    }
`;

export const PAGINATION_QUERY = gql`
    query PAGINATION_QUERY {
        itemsConnection {
            aggregate {
                count
            }
        }
    }
`;

export const SIGNUP_MUTATION = gql`
    mutation SIGNUP_MUTATION(
    $email: String!
    $name: String!
    $password: String!
    ) {
        signup(
            email: $email
            name: $name
            password: $password
        ) {
            id
            email
            name
        }
    }
`;

export const SIGNIN_MUTATION = gql`
    mutation SIGNIN_MUTATION(
    $email: String!
    $password: String!
    ) {
        signin(
            email: $email
            password: $password
        ) {
            id
            email
            name
        }
    }
`;

export const CURRENT_USER_QUERY = gql`
    query CURRENT_USER_QUERY {
        currentUser {
            id
            email
            name
            permissions
			orders {
				id
			}
            cart {
                id
                quantity
                item {
                    id
                    price
                    image
                    title
                    description
                }
            }
        }
    }
`;

export const SIGNOUT_MUTATION = gql`
    mutation SIGNOUT_MUTATION {
        signout {
            message
        }
    }
`;

export const RESET_PASSWORD_MUTATION = gql`
    mutation RESET_PASSWORD_MUTATION(
    $password: String!
    $confirmPassword: String!
    $resetToken: String!
    ) {
        resetPassword(
            password: $password
            confirmPassword: $confirmPassword
            resetToken: $resetToken
        ) {
            id
            email
            name
        }
    }
`;

export const REQUEST_RESET_MUTATION = gql`
    mutation REQUEST_RESET_MUTATION(
    $email: String!
    ) {
        requestReset(
            email: $email
        ) {
            message
        }
    }
`;

export const ALL_USERS_QUERY = gql`
    query ALL_USERS_QUERY {
        users {
            id
            name
            email
            permissions
        }
    }
`;

export const UPDATE_PERMISSIONS_MUTATION = gql`
    mutation UPDATE_PERMISSIONS_MUTATION(
    $permissions: [Permission]
    $userId: ID!
    ) {
        updatePermissions(
            permissions: $permissions
            userId: $userId
        ) {
            id
            name
            email
            permissions
        }
    }
`;

// @client tells Apollo to not go to the GraphQL client or the server but check the client state for cartOpen
export const LOCAL_STATE_QUERY = gql`
    query LOCAL_STATE_QUERY {
        cartOpen @client
    }
`;

export const TOGGLE_CART_MUTATION = gql`
    mutation TOGGLE_CART_MUTATION {
        toggleCart @client
    }
`;

export const ADD_TO_CART_MUTATION = gql`
    mutation ADD_TO_CART_MUTATION(
    $id: ID!
    ) {
        addToCart(
            id: $id
        ) {
            id
            quantity
        }
    }
`;

export const REMOVE_FROM_CART_MUTATION = gql`
    mutation REMOVE_FROM_CART_MUTATION(
    $id: ID!
    ) {
        removeFromCart(
            id: $id
        ) {
            id
        }
    }
`;

export const SEARCH_ITEMS_QUERY = gql`
    query SEARCH_ITEMS_QUERY(
    $searchTerm: String!
    ) {
        items(
            where: {
                OR: [
                    { title_contains: $searchTerm },
                    { description_contains: $searchTerm },
                ],

            }
        ) {
			id
			image
			title
		}
    }
`;

export const CREATE_ORDER_MUTATION = gql`
	mutation CREATE_ORDER_MUTATION(
		$token: String!
	) {
		createOrder(
			token: $token
		) {
			id 
			charge
			total
			items {
				id 
				title
			}
		}
	}
`;

export const ORDER_QUERY = gql`
	query ORDER_QUERY (
    $id: ID!
    ) {
		order(
			id: $id
		) {
			id
			items {
				id 
				title
				description
				price
				image
				quantity
			}
			total
			user {
				id
			}
			charge
			createdAt
		}
	}
`;

export const ORDERS_QUERY = gql`
	query ORDERS_QUERY {
		orders(
			OrderBy: createdAt_DESC
		) {
			id
			total
			createdAt
			items {
				id
				title
				price
				description
				quantity
				image
			}
		}
	}
`;