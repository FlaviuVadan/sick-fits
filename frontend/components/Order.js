import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import { format } from 'date-fns';
// allows to change title
import Head from 'next/head';
import formatMoney from '../lib/formatMoney';
import Error from './ErrorMessage';
import OrderStyles from './styles/OrderItemStyles';
import { ORDER_QUERY } from "../queries/queries";


class Order extends Component {
	static propTypes = {
		id: PropTypes.string.isRequired,
	};

	render() {
		return (
			<Query query={ORDER_QUERY} variables={{
				id: this.props.id,
			}}>
				{({ data, error, loading }) => {
					if (error) return <Error error={error}/>;
					if (loading) return <p>Loading...</p>;
					console.log(data)
					return (
						<div>
							<p>Order ID: {this.props.id}</p>
						</div>
					);
				}}
			</Query>
		);
	}
}

export default Order;