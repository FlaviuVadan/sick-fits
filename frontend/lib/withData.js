// a high order component that exposes the Apollo client via a prop
// want server side rendering and next-with-apollo helps with that
import withApollo from 'next-with-apollo';

// package by Apollo containing all the std things one may use
// can extend what the Apollo client does and contains a lot of pre-configed things
import ApolloClient from 'apollo-boost';
import { endpoint } from '../config';

import { LOCAL_STATE_QUERY } from "../queries/queries";


// header important for authentication
function createClient({ headers }) {
	return new ApolloClient({
		uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
		// include credentials on every request, includes cookies
		request: operation => {
			operation.setContext({
				fetchOptions: {
					credentials: 'include',
				},
				headers,
			});
		},
		// local data
		clientState: {
			resolvers: {
				Mutation: {
					// toggleCart will have one variable garbage collected
					toggleCart(_, variables, { cache }) {
						// read cartOpen value from cache
						const { cartOpen } = cache.readQuery({
							query: LOCAL_STATE_QUERY,
						});
						// write cart state to opposite
						const data = {
							data: {
								cartOpen: !cartOpen,
							},
						};
						cache.writeData(data);
						return data;
					},
				},
			},
			defaults: {
				cartOpen: true
			}
		}
	});
}

export default withApollo(createClient);
