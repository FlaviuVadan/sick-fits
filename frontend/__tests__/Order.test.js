import { mount } from 'enzyme';
import Router from 'next/router';
import NProgress from 'nprogress';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import Order from '../components/Order';
import { ORDER_QUERY } from "../queries/queries";
import { fakeOrder } from "../lib/testUtils";

const mocks = [
	{
		request: {
			query: ORDER_QUERY,
			variables: {
				id: "ord123",
			},
		},
		result: {
			data: {
				order: {
					...fakeOrder(),
				},
			},
		},
	},
];

describe('<Order/>', () => {
	it('renders', async () => {
		const wrapper = mount(
			<MockedProvider mocks={mocks}>
				<Order id="ord123"/>
			</MockedProvider>
		);
		await wait();
		wrapper.update();
		const order = wrapper.find('[data-test="order"]');

		// the snapshot also checks values on the page
		expect(toJSON(order)).toMatchSnapshot();
	});
});
