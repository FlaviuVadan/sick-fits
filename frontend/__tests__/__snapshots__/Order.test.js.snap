// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order/> renders 1`] = `
Array [
  <OrderItemStyles
    data-test="order"
  >
    <li
      className="OrderItemStyles-fbwo89-0 jkaGkx"
      data-test="order"
    >
      <SideEffect(Head)>
        <Head />
      </SideEffect(Head)>
      <p>
        <span>
          Order ID: 
        </span>
        <span>
          ord123
        </span>
      </p>
      <p>
        <span>
          Charge: 
        </span>
        <span>
          ch_123
        </span>
      </p>
      <p>
        <span>
          Date: 
        </span>
        <span>
          March 31, 2018 6:00
        </span>
      </p>
      <p>
        <span>
          Order total: 
        </span>
        <span>
          $400
        </span>
      </p>
      <p>
        <span>
          Item count: 
        </span>
        <span>
          2
        </span>
      </p>
      <div
        className="items"
      >
        <div
          className="order-item"
          key="3a430a73-c9e9-4bcc-b46a-41618965ffea"
        >
          <img
            alt="soluta non omnis consequatur enim quia autem"
            src="reprehenderit.jpg"
          />
          <div
            className="item-details"
          >
            <h2>
              soluta non omnis consequatur enim quia autem
            </h2>
            <p>
              Description: 
              et modi tenetur amet modi reprehenderit omnis
            </p>
            <p>
              Qty: 
              1
            </p>
            <p>
              Each: 
              $42.34
            </p>
            <p>
              Subtotal: 
              $42.34
            </p>
          </div>
        </div>
        <div
          className="order-item"
          key="2393e090-592f-4d03-b808-c9d81098deec"
        >
          <img
            alt="quia sed exercitationem omnis laborum exercitationem est"
            src="sint.jpg"
          />
          <div
            className="item-details"
          >
            <h2>
              quia sed exercitationem omnis laborum exercitationem est
            </h2>
            <p>
              Description: 
              sapiente laudantium molestias assumenda quasi adipisci mollitia
            </p>
            <p>
              Qty: 
              1
            </p>
            <p>
              Each: 
              $42.34
            </p>
            <p>
              Subtotal: 
              $42.34
            </p>
          </div>
        </div>
      </div>
    </li>
  </OrderItemStyles>,
  <li
    className="OrderItemStyles-fbwo89-0 jkaGkx"
    data-test="order"
  >
    <SideEffect(Head)>
      <Head />
    </SideEffect(Head)>
    <p>
      <span>
        Order ID: 
      </span>
      <span>
        ord123
      </span>
    </p>
    <p>
      <span>
        Charge: 
      </span>
      <span>
        ch_123
      </span>
    </p>
    <p>
      <span>
        Date: 
      </span>
      <span>
        March 31, 2018 6:00
      </span>
    </p>
    <p>
      <span>
        Order total: 
      </span>
      <span>
        $400
      </span>
    </p>
    <p>
      <span>
        Item count: 
      </span>
      <span>
        2
      </span>
    </p>
    <div
      className="items"
    >
      <div
        className="order-item"
        key="3a430a73-c9e9-4bcc-b46a-41618965ffea"
      >
        <img
          alt="soluta non omnis consequatur enim quia autem"
          src="reprehenderit.jpg"
        />
        <div
          className="item-details"
        >
          <h2>
            soluta non omnis consequatur enim quia autem
          </h2>
          <p>
            Description: 
            et modi tenetur amet modi reprehenderit omnis
          </p>
          <p>
            Qty: 
            1
          </p>
          <p>
            Each: 
            $42.34
          </p>
          <p>
            Subtotal: 
            $42.34
          </p>
        </div>
      </div>
      <div
        className="order-item"
        key="2393e090-592f-4d03-b808-c9d81098deec"
      >
        <img
          alt="quia sed exercitationem omnis laborum exercitationem est"
          src="sint.jpg"
        />
        <div
          className="item-details"
        >
          <h2>
            quia sed exercitationem omnis laborum exercitationem est
          </h2>
          <p>
            Description: 
            sapiente laudantium molestias assumenda quasi adipisci mollitia
          </p>
          <p>
            Qty: 
            1
          </p>
          <p>
            Each: 
            $42.34
          </p>
          <p>
            Subtotal: 
            $42.34
          </p>
        </div>
      </div>
    </div>
  </li>,
]
`;
